modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -641,7 +641,23 @@ def to_pandas(cls, partitions):
         pandas.DataFrame
             A pandas DataFrame
         """
-        retrieved_objects = [[obj.to_pandas() for obj in part] for part in partitions]
+        retrieved_objects = cls.get_objects_from_partitions(partitions.flatten())
+        if all(
+            map(
+                lambda obj: isinstance(obj, (pandas.DataFrame, pandas.Series)),
+                retrieved_objects,
+            )
+        ):
+            old_shape = partitions.shape
+            # restore 2d array
+            retrieved_objects = [
+                [retrieved_objects[i * old_shape[1] + j] for j in range(old_shape[1])]
+                for i in range(old_shape[0])
+            ]
+        else:
+            retrieved_objects = [
+                [obj.to_pandas() for obj in part] for part in partitions
+            ]
```

Comment:
isn't this doing a double job for line 644?

Reply:
that would make an awesome in-code comment

Reply:
added